### PR TITLE
Recognize identifier characters from the Supplementary Multilingual Plane

### LINKED ISSUE #1244: Recognize identifier characters from the Supplementary Multilingual Plane
This is the logical next step after #1232. The characters from SMP, either as IdentifierStart of IdentifierPart, encoded via Unicode code point escapes or surrogate pairs, need to be recognized properly.

### PR DESCRIPTION
**Pull request by ariya**
Fixes #1244.

### CODE REVIEW

**Comment by ariya** [comment]
Ugh, I missed some additional test files. Stay tuned...

**Comment by ariya** [comment]
The failure on Travis CI is due to the failing downstream tests with JSCS.

@mikesherov  Seems that JSCS master doesn't really pass, i.e. the following:

```
git clone https://github.com/jscs-dev/node-jscs.git
cd node-jscs
npm install
npm test
```

fails for me (Node.js v0.12.0, npm 2.9.1).

**Comment by mikesherov** [comment]
Let's skip jscs master for now.

**Comment by ariya** [comment]
@mikesherov I'll commit something quickly on master to skip JSCS, so that any PRs will be unblocked.

**Comment by ariya** [comment]
@ikarienator @mikesherov @michaelficarra @mathiasbynens Anyone willing to review this?

File: esprima.js [hunk 1]
```diff
@@ -605,23 +611,31 @@
...
+
+    function codePointAt(i) {
+        var cp, first, second;
+
+        cp = source.charCodeAt(i);
+        if (cp >= 0xD800 && cp <= 0xDBFF) {
+            first = cp;
+            second = source.charCodeAt(i + 1);
+            cp = (first - 0xD800) * 0x400 + second - 0xDC00 + 0x10000;
         }
```
Comment (mathiasbynens) [inline esprima.js]: What about lone surrogates? E.g. `\uD800` followed by `x`.

Comment (ariya) [inline esprima.js]: Good catch! I'll make sure that will trigger an error.

**Comment by mikesherov** [comment]
I completely defer to @mathiasbynens's expertise in this area.

**Comment by ariya** [comment]
Anything else I should worry about before I am checking this in?

**Comment by mathiasbynens** [comment]
I collected some edge cases that might make for interesting tests here: https://mathiasbynens.be/notes/javascript-identifiers-es6 But the patch looks good to me as-is.

**Comment by ariya** [comment]
@mathiasbynens I think I've covered most, if not all, of the corner cases you've described. Very helpful, thank you so much for that!
